README.md
# Factorio mod : Spawner alert

This mod alerts you when spawners start consuming pollution. Surprise them before they surprise you!

Default binding to enable/disable alert is : __SHIFT + I__.


changelog.txt
---------------------------------------------------------------------------------------------------
Version: 0.1.8
Date: 2021-03-07
  Changes:
    - Update mod for factorio version 1.1.25.
---------------------------------------------------------------------------------------------------
Version: 0.1.7
Date: 2020-11-29
  Changes:
    - Update mod for factorio version 1.1.
---------------------------------------------------------------------------------------------------
Version: 0.1.6
Date: 2020-09-26
  Bugfixes:
    - Fixed a bug where mod crashes if alerts were not toggled.
---------------------------------------------------------------------------------------------------
Version: 0.1.5
Date: 2020-09-19
  Bugfixes:
    - Fixed a bug where mod would crash at startup if pollution consumption was greater than 0.
---------------------------------------------------------------------------------------------------
Version: 0.1.4
Date: 2020-09-19
  Bugfixes:
    - Fixed a bug where mod would crash when player enables alerts.
---------------------------------------------------------------------------------------------------
Version: 0.1.3
Date: 2020-09-18
  Bugfixes:
    - Fixed a bug where the player gets desynchronized in a multiplayer game at alert enable.
---------------------------------------------------------------------------------------------------
Version: 0.1.2
Date: 2020-09-12
  Bugfixes:
    - Fixed a bug where alerts where not shown when consummed pollution was low.
    - Fixed a bug where mod would crash in god mode.
---------------------------------------------------------------------------------------------------
Version: 0.1.1
Date: 2020-08-29
  Changes:
    - Updated alert icon.
---------------------------------------------------------------------------------------------------
Version: 0.1.0
Date: 2020-08-27
  Info:
    - Initial release.

control.lua
--
-- Factorio: Spawner alert mod
-- File: control.lua
-- Date: 25/08/2020 23:49:03
-- Author: Ish (persoishtar@gmail.com)
-- --------------------------------------
-- http://www.opensource.org/licenses/MIT
-- MIT License
-- 
-- Copyright (c) 2020 Ish
-- 
-- Permission is hereby granted, free of charge, to any person obtaining a copy of
-- this software and associated documentation files (the "Software"), to deal in
-- the Software without restriction, including without limitation the rights to
-- use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
-- of the Software, and to permit persons to whom the Software is furnished to do
-- so, subject to the following conditions:
-- 
-- The above copyright notice and this permission notice shall be included in all
-- copies or substantial portions of the Software.
-- 
-- THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
-- IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
-- FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
-- AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
-- LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
-- OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
-- SOFTWARE.
--
local input_control_name = "spawner-alert-toggle"

-- ------------------------------------------------------------------------------------ --
--                                       Handlers                                       --
-- ------------------------------------------------------------------------------------ --

function toggle_alert(event)
    local player = game.get_player(event.player_index)

    -- toggle alert switch
    if (global.players[player.index] == nil) then
        global.players[player.index] = {alert = true}
    else
        global.players[player.index].alert = not (global.players[player.index].alert)
    end

    player.print(
        "Spawner alert status: " ..
            (global.players[player.index].alert and 'enabled' or 'disabled'),
        {r = 1, g = 0.25, b = 0.25}
    )
end

function check_and_alert(event)
    -- compute total number of pollution consumption by spawners
    local stats_to_fetch = {
        name = "spitter-spawner",
        input = false,
        precision_index = 
            defines.flow_precision_index.one_second or 
            defines.flow_precision_index.five_seconds,
        count = true
    }

    local spitter_count = game.pollution_statistics.get_flow_count(stats_to_fetch)

    stats_to_fetch.name = "biter-spawner"
    local biter_count = game.pollution_statistics.get_flow_count(stats_to_fetch)
    local pollution_count = math.ceil(spitter_count + biter_count)

    -- alert players if alert is enabled
    if pollution_count > 0 then
        local players = game.players

        for _, player in pairs(players) do
            if (player.character and global.players[player.index].alert) then
                player.add_custom_alert(
                    player.character, {type = "fluid", name = "spawner-alert-icon"},
                    "Spawners are consuming " .. pollution_count .. " pollution/s.", true
                )
            end
        end
    end
end

function player_joined(player)
    index = player.index or player.player_index
    global.players[index] = global.players[index] or {}
    global.players[index].alert = global.players[index].alert or true
end

function init_mod(event)
    -- initialize global players variable
    global.players = global.players or {}

    -- initialize each player
    for _, player in pairs(game.players) do
        player_joined(player)
    end
end

-- ------------------------------------------------------------------------------------ --
--                                        Events                                        --
-- ------------------------------------------------------------------------------------ --

script.on_init(init_mod)
script.on_configuration_changed(init_mod)
script.on_event(defines.events.on_player_joined_game, player_joined)
script.on_event(input_control_name, toggle_alert)
script.on_nth_tick(60, check_and_alert)


prototypes/signals.lua
--
-- Factorio: Spawner alert mod
-- File: signals.lua
-- Date: 29/08/2020 18:59:22
-- Author: Ish (persoishtar@gmail.com)
-- --------------------------------------
-- http://www.opensource.org/licenses/MIT
-- MIT License
-- 
-- Copyright (c) 2020 Ish
-- 
-- Permission is hereby granted, free of charge, to any person obtaining a copy of
-- this software and associated documentation files (the "Software"), to deal in
-- the Software without restriction, including without limitation the rights to
-- use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
-- of the Software, and to permit persons to whom the Software is furnished to do
-- so, subject to the following conditions:
-- 
-- The above copyright notice and this permission notice shall be included in all
-- copies or substantial portions of the Software.
-- 
-- THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
-- IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
-- FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
-- AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
-- LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
-- OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
-- SOFTWARE.
--

data:extend(
    {
        {
            type = "item-subgroup",
            name = "spawner-alert-signal",
            group = "signals"
        },
        {
            type = "fluid",
            name = "spawner-alert-icon",
            icon = "__spawner-alert__/graphics/icons/biter_nest.png",
            icon_size = 128,
            icon_mipmaps = 0,
            default_temperature = 15,
            max_temperature = 100,
            heat_capacity = "0.2KJ",
            base_color = {r = 0, g = 0, b = 0},
            flow_color = {r = 0, g = 0, b = 0},
            hidden = true,
            auto_barrel = false,
            subgroup = "spawner-alert-signal"
        }
    }
)
